### PR TITLE
Adds new metric for dropped samples in ingester

### LINKED ISSUE #3954: Add a metric to track samples dropped in distributor by metric_relabel
If you have a `metric_relabel_configs ... action: drop` then those samples disappear from `cortex_distributor_received_samples_total`. 

I'd like them to be counted in `cortex_discarded_samples_total`.

Relating to the feature added in #3329

### PR DESCRIPTION
**Pull request by pedro-stanaka**
**What this PR does**: As described in #3954, this PR adds a new metric `distributor_discarded_samples_total` to represent 
the number of dropped samples due to relabeling configuration.

**Which issue(s) this PR fixes**: 
Fixes #3954

**Checklist**
- [X] Tests updated
- [x] Documentation added (N/A ?)
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: pkg/distributor/distributor.go [hunk 1]
```diff
@@ -250,6 +251,11 @@ func New(cfg Config, clientConfig ingester_client.Config, limits *validation.Ove
 			Name:      "distributor_received_samples_total",
 			Help:      "The total number of received samples, excluding rejected and deduped samples.",
 		}, []string{"user"}),
+		discardedSamples: promauto.With(reg).NewCounterVec(prometheus.CounterOpts{
```
Comment (alvinlin123) [inline pkg/distributor/distributor.go:254]: I think we should re-use `cortex_discarded_samples_total` defined at https://github.com/cortexproject/cortex/blob/34bb142b129bd781539001fb1629cf2f3ed54b82/pkg/util/validation/validate.go#L66

We can simply add a new reason string.

Comment (pedro-stanaka) [reply pkg/distributor/distributor.go]: Nice one, I did as suggested. I didn't know about that metric.

File: pkg/distributor/distributor.go [hunk 2]
```diff
@@ -637,6 +637,13 @@ func (d *Distributor) Push(ctx context.Context, req *cortexpb.WriteRequest) (*co
 
 		if mrc := d.limits.MetricRelabelConfigs(userID); len(mrc) > 0 {
 			l := relabel.Process(cortexpb.FromLabelAdaptersToLabels(ts.Labels), mrc...)
+			if l == nil {
+				// all labels are gone, therefore the __name__ label is not present, metric will be discarded
```
Comment (bboreham) [inline pkg/distributor/distributor.go:641]: A better test for "all labels are gone" would be `len(l) == 0`.  I haven't investigated whether the code or the comment should be changed.

File: pkg/distributor/distributor.go [hunk 3]
```diff
@@ -637,6 +637,13 @@ func (d *Distributor) Push(ctx context.Context, req *cortexpb.WriteRequest) (*co
...
+				validation.DiscardedSamples.WithLabelValues(
+					validation.DroppedByRelabelConfiguration,
+					userID,
+				).Add(float64(len(ts.Samples)))
+			}
```
Comment (bboreham) [inline pkg/distributor/distributor.go:646]: `continue` here is clearer, I think, rather than relying on a test later on.

File: pkg/distributor/distributor.go [hunk 4]
```diff
@@ -707,7 +714,7 @@ func (d *Distributor) Push(ctx context.Context, req *cortexpb.WriteRequest) (*co
 	}
 
 	d.receivedSamples.WithLabelValues(userID).Add(float64(validatedSamples))
-	d.receivedExemplars.WithLabelValues(userID).Add((float64(validatedExemplars)))
+	d.receivedExemplars.WithLabelValues(userID).Add(float64(validatedExemplars))
```
Comment (bboreham) [inline pkg/distributor/distributor.go:717]: Ideally, don't make unrelated changes.

File: pkg/distributor/distributor_test.go [hunk 5]
```diff
@@ -2654,30 +2655,37 @@ func TestSortLabels(t *testing.T) {
 }
 
 func TestDistributor_Push_Relabel(t *testing.T) {
-	ctx := user.InjectOrgID(context.Background(), "user")
+	const orgID = "user"
+	ctx := user.InjectOrgID(context.Background(), orgID)
```
Comment (bboreham) [inline pkg/distributor/distributor_test.go:2659]: why was this change made?

File: pkg/distributor/distributor_test.go [hunk 6]
```diff
@@ -1194,16 +1194,16 @@ func TestDistributor_Push_LabelRemoval(t *testing.T) {
 func TestDistributor_Push_ShouldGuaranteeShardingTokenConsistencyOverTheTime(t *testing.T) {
 	ctx := user.InjectOrgID(context.Background(), "user")
 	tests := map[string]struct {
-		inputSeries    labels.Labels
+		inputSeries    []labels.Labels
```
Comment (bboreham) [inline pkg/distributor/distributor_test.go:1197]: I think most of the tests where you made this change should be done the same way `TestDistributor_MetricsForLabelMatchers` was, to minimise diffs and reviewer effort.

File: pkg/util/validation/validate.go [hunk 7]
```diff
@@ -34,16 +34,16 @@ const (
...
+	missingMetricName             = "missing_metric_name"
+	invalidMetricName             = "metric_name_invalid"
+	greaterThanMaxSampleAge       = "greater_than_max_sample_age"
+	maxLabelNamesPerSeries        = "max_label_names_per_series"
+	tooFarInFuture                = "too_far_in_future"
+	invalidLabel                  = "label_invalid"
+	labelNameTooLong              = "label_name_too_long"
+	duplicateLabelNames           = "duplicate_label_names"
+	labelsNotSorted               = "labels_not_sorted"
+	labelValueTooLong             = "label_value_too_long"
```
Comment (bboreham) [inline pkg/util/validation/validate.go:46]: Why did the formatting change here?

**Review by bboreham — COMMENTED** [review]
Thanks for the PR; I think there are a number of unnecessary changes, or if they are necessary they should go in a different PR.

A good cross-check is to make sure all changes are explained in the commit message.

File: pkg/distributor/distributor.go (hunk 4 above)
Comment (pedro-stanaka) [reply pkg/distributor/distributor.go]: Reverted it. Just was a low-hanging fruit and didn't see evil, but I get it generates noise.

File: pkg/distributor/distributor.go (hunk 3 above)
Comment (pedro-stanaka) [reply pkg/distributor/distributor.go:647]: Done, pulled the continue to here.

File: pkg/distributor/distributor.go (hunk 2 above)
Comment (pedro-stanaka) [reply pkg/distributor/distributor.go]: Yeah, the underlying code from `relabel.Process` returns nil (so I thought it was okay to compare with nil), but `len` will catch empty slices as well, so I think it is better.

File: pkg/distributor/distributor_test.go (hunk 5 above)
Comment (pedro-stanaka) [reply pkg/distributor/distributor_test.go]: I was using this const down the test in during an exploration of mine, forgot to undo this change. Now it is undone.

File: pkg/distributor/distributor_test.go (hunk 6 above)
Comment (pedro-stanaka) [reply pkg/distributor/distributor_test.go]: Done, you are completely right.

File: pkg/util/validation/validate.go (hunk 7 above)
Comment (pedro-stanaka) [reply pkg/util/validation/validate.go:46]: Hum, there I can't really explain, I didnt change that myself so maybe it was a gofmt hook, but I see no changes for me, the old blame is kept:

image

**Comment by pedro-stanaka** [comment]
@bboreham @alvinlin123 anything else I should be doing here? Can I have further feedback on the PR?

**Comment by pedro-stanaka** [comment]
Lint output:
image

test output:
image

File: pkg/distributor/distributor.go (hunk 3 above)
Comment (alvinlin123) [reply pkg/distributor/distributor.go:647]: But at line 650/658 we have 

```
for _, labelName := range d.limits.DropLabels(userID) {
  removeLabel(labelName, &ts.Labels)
}
```
Assuming we don't have relabel config, but we configure Cortex to remove the  `__name__` label (and assuming the series only have `__name__` label), then we will end up with empty labels. So we should still do another `len(ts.Labels) == 0` after processing the `drop_labels` configuration right?

Comment (pedro-stanaka) [reply pkg/distributor/distributor.go:647]: I think this will never happen because if for a certain user we drop `__name__` label this means that for that user we won't ingest any metrics. But I added this if again and also created a test case to cover this scenario. I had to slightly modify the if to check for the `__name__` label as well.

File: pkg/distributor/distributor.go [hunk 8]
```diff
@@ -637,6 +637,14 @@ func (d *Distributor) Push(ctx context.Context, req *cortexpb.WriteRequest) (*co
 
 		if mrc := d.limits.MetricRelabelConfigs(userID); len(mrc) > 0 {
 			l := relabel.Process(cortexpb.FromLabelAdaptersToLabels(ts.Labels), mrc...)
+			if len(l) == 0 {
+				// all labels are gone, therefore the __name__ label is not present, samples will be discarded
```
Comment (bboreham) [inline pkg/distributor/distributor.go:641]: remove mention of `__name__`

File: pkg/distributor/distributor.go [hunk 9]
```diff
@@ -651,7 +659,12 @@ func (d *Distributor) Push(ctx context.Context, req *cortexpb.WriteRequest) (*co
 			removeLabel(labelName, &ts.Labels)
 		}
 
-		if len(ts.Labels) == 0 {
+		if len(ts.Labels) == 0 || wasNameLabelRemoved(ts.Labels) {
```
Comment (bboreham) [inline pkg/distributor/distributor.go:662]: Don't look at `__name__`.

**Review by bboreham — COMMENTED** [review]
Sorry I let this languish for a while.
Insisting on `__name__` is optional - https://github.com/cortexproject/cortex/blob/4e9fc3a2b5ab6a76787d59eb441f3af2765f1de8/pkg/util/validation/limits.go#L137
so please don't add more checks for `__name__`.  Just check whether any labels remain.

File: pkg/distributor/distributor.go (hunk 9 above)
Comment (pedro-stanaka) [reply pkg/distributor/distributor.go]: Sure thing, done.

File: pkg/distributor/distributor.go (hunk 8 above)
Comment (pedro-stanaka) [reply pkg/distributor/distributor.go]: Done.

**Comment by pedro-stanaka** [comment]
I changed as suggested, I think this is finally good to go. Please let me know if I should do anything else. On my machine, linting and tests are passing.

**Comment by pedro-stanaka** [comment]
@bboreham sorry for the endless ping, but I just want to make sure you get notifications. I implemented your last comments a while ago, checking back to see if this is good to go.

**Review by bboreham — APPROVED** [review]
LGTM, thanks!

**Comment by pedro-stanaka** [comment]
@alvinlin123 can you please recheck as well so we can merge this? I am going to do yet another rebase.

**Review by alvinlin123 — APPROVED** [review]
